File: src/main/java/core/procedures/exceptions/ExInfo.java
package core.procedures.exceptions;

import core.exceptions.ExInfoException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.Map;

public class ExInfo extends AFn {

  public ExInfo() {
    super(new FnArgsBuilder().min(2).max(3).mandatory(new Class[]{String.class, Map.class}).rest(Throwable.class).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "ex-info";
  }

  @Override
  public ExInfoException apply(Object... args) {
    if (args.length == 2) {
      return new ExInfoException((String)args[0], (Map)args[1]);
    }
    return new ExInfoException((String)args[0], (Map)args[1], (Throwable)args[2]);
  }
}

File: src/main/java/core/procedures/exceptions/ExInfo.kt
package core.procedures.exceptions

import core.exceptions.ExInfoException
import core.procedures.AFn
import core.procedures.Arity.Range

class ExInfo : AFn<Any?, ExInfoException>(name = "ex-info", isPure = true, arity = Range(2, 3),
                                          mandatoryArgsTypes = arrayOf(String::class.java, Map::class.java),
                                          restArgsType = Throwable::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        2    -> ExInfoException(args[0] as String, args[1] as Map<*, *>)
        else -> ExInfoException(args[0] as String, args[1] as Map<*, *>, args[2] as Throwable)
    }
}
